### PR TITLE
[RUM] Adds RUM Application resource support

### PR DESCRIPTION
**Pull request by nkzou**
Continues #1526 , adds cassettes and some doc improvements

### CODE REVIEW

**Review by benoitmeriaux — APPROVED** [review]

**Comment by benoitmeriaux** [comment]
I'm not very familiar with terraform provider, but it looks good :)

File: datadog/resource_datadog_rum_application.go [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+import (
+	"context"
+
+	"github.com/DataDog/datadog-api-client-go/v2/api/datadog"
+	"github.com/DataDog/datadog-api-client-go/v2/api/datadogV2"
+	"github.com/terraform-providers/terraform-provider-datadog/datadog/internal/utils"
+
+	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+)
```
Comment (skarimo) [inline datadog/resource_datadog_rum_application.go:12]:
```suggestion
import (
	"context"

	"github.com/terraform-providers/terraform-provider-datadog/datadog/internal/utils"

	"github.com/DataDog/datadog-api-client-go/v2/api/datadog"
	"github.com/DataDog/datadog-api-client-go/v2/api/datadogV2"
	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
)
```

File: datadog/resource_datadog_rum_application.go [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+				Type:        schema.TypeString,
+				Optional:    true,
+				Default:     "browser",
+				Description: "The RUM application type. Supported values are `browser`, `ios`, `android`, `react-native`, `flutter`",
+			},
+			"client_token": {
+				Type:        schema.TypeString,
+				Computed:    true,
+				Description: "The RUM application client token",
+			},
```
Comment (skarimo) [inline datadog/resource_datadog_rum_application.go:40]: Should we add the computed field `application_id` in the schema as well? Looks like it needs to be used with `client_token`. Or is this same as ID?

File: datadog/resource_datadog_rum_application.go [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+	}
+}
+
+func resourceDatadogRUMApplicationRead(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	providerConf := meta.(*ProviderConfiguration)
+	apiInstances := providerConf.DatadogApiInstances
+	auth := providerConf.Auth
+
+	resp, httpResponse, err := apiInstances.GetRumApiV2().GetRUMApplication(auth, d.Id())
+
```
Comment (skarimo) [inline datadog/resource_datadog_rum_application.go:51]:
```suggestion
```

File: datadog/resource_datadog_rum_application.go [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+			},
+			Type: datadogV2.RUMAPPLICATIONCREATETYPE_RUM_APPLICATION_CREATE,
+		},
+	}
+
+	providerConf := meta.(*ProviderConfiguration)
+	apiInstances := providerConf.DatadogApiInstances
+	auth := providerConf.Auth
+	resp, httpResponse, err := apiInstances.GetRumApiV2().CreateRUMApplication(auth, body)
+
```
Comment (skarimo) [inline datadog/resource_datadog_rum_application.go:79]:
```suggestion
```

File: datadog/resource_datadog_rum_application.go [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+			Id:   d.Id(),
+			Type: datadogV2.RUMAPPLICATIONUPDATETYPE_RUM_APPLICATION_UPDATE,
+		},
+	}
+
+	providerConf := meta.(*ProviderConfiguration)
+	apiInstances := providerConf.DatadogApiInstances
+	auth := providerConf.Auth
+	resp, httpResponse, err := apiInstances.GetRumApiV2().UpdateRUMApplication(auth, d.Id(), body)
+
```
Comment (skarimo) [inline datadog/resource_datadog_rum_application.go:105]:
```suggestion
```

File: datadog/resource_datadog_rum_application.go [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+	return updateRUMApplicationState(d, resp.Data)
+}
+
+func resourceDatadogRUMApplicationDelete(_ context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	providerConf := meta.(*ProviderConfiguration)
+	apiInstances := providerConf.DatadogApiInstances
+	auth := providerConf.Auth
+
+	httpResponse, err := apiInstances.GetRumApiV2().DeleteRUMApplication(auth, d.Id())
+
```
Comment (skarimo) [inline datadog/resource_datadog_rum_application.go:119]:
```suggestion
```

File: datadog/resource_datadog_rum_application.go (hunk 2 above)
Comment (buranmert) [reply datadog/resource_datadog_rum_application.go:40]: i think `"The RUM application client token"` is sort of confusing: `RUM Application ID` and `Client Token` are 2 different things and both are required

Only `"The client token"` as description should be better.

File: docs/resources/rum_application.md [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+
+- `name` (String) The name of the RUM application
+
+### Optional
+
+- `type` (String) The RUM application type. Supported values are `browser`, `ios`, `android`, `react-native`, `flutter`
+
+### Read-Only
+
+- `client_token` (String) The RUM application client token
```
Comment (buranmert) [inline docs/resources/rum_application.md:35]:
```suggestion
- `client_token` (String) The client token
```

**Review by buranmert — COMMENTED** [review]
thanks for implementing this 🙇 

i left a comment regarding the potential confusion around "RUM Application ID" vs "Client Token"

**Comment by SebastianBalle** [comment]
@nkzou is it possible for you to implement the data resource for the rum_application? We share a single RUM application in multiple environments but manages the rum application resource in one. We would however like to know the client_token and id in other environments.

**Pull request merged**